### PR TITLE
Update `asyncio.tasks` for py312

### PR DESCRIPTION
**Pull request by AlexWaygood**
Runtime implementation added in https://github.com/python/cpython/commit/a474e04388c2ef6aca75c26cb70a1b6200235feb

### CODE REVIEW

File: stdlib/asyncio/tasks.pyi [hunk 1]
```diff
@@ -356,5 +359,40 @@ else:
...
+
+    class _CustomTaskConstructor(Protocol[_TaskT_co]):
+        def __call__(
+            self,
+            __coro: _TaskCompatibleCoro[Any],
+            loop: AbstractEventLoop,
+            name: str | None,
+            context: Context | None,
+            eager_start: bool,
+        ) -> _TaskT_co: ...
```
Comment (AlexWaygood) [inline stdlib/asyncio/tasks.pyi:374]: I think the idea might be that it's permissible for the custom task constructor to return anything that _duck-types_ as an `asyncio.Task` -- it doesn't necessarily _have_ to be a subclass of `asyncio.Task`. That would be even more complicated than what we have currently, however... I'd vote for keeping it simple until we get a complaint that we're not allowing somebody's duck-type-compatible third-party `Task` implementation

File: stdlib/asyncio/tasks.pyi [hunk 2]
```diff
@@ -356,5 +381,40 @@ else:
...
 def _leave_task(loop: AbstractEventLoop, task: Task[Any]) -> None: ...
+
+if sys.version_info >= (3, 12):
+    _TaskT_co = TypeVar("_TaskT_co", bound=Task[Any], covariant=True)
+
+    class _CustomTaskConstructor(Protocol[_TaskT_co]):
+        def __call__(
+            self,
+            __coro: _TaskCompatibleCoro[Any],
+            loop: AbstractEventLoop,
```
Comment (srittau) [inline stdlib/asyncio/tasks.pyi:392]: To make it a bit more lenient:

```suggestion
            __coro: _TaskCompatibleCoro[Any],
            *,
            loop: AbstractEventLoop,
```

**Review by srittau — APPROVED** [review]
One nit, but looks good.

File: stdlib/asyncio/tasks.pyi (hunk 2 above)
Comment (AlexWaygood) [reply stdlib/asyncio/tasks.pyi:392]: nice catch!

**Pull request merged**